jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomJtaPlatform.java

```diff
@@ -22,39 +22,61 @@
+    // is reset every time the Transaction Manager service is restarted,
+    // as (application deployment) JPA persistence unit service depends on the TM service.
+    // For this reason, the static 'transactionSynchronizationRegistry' can be updated.
+    // Note that Hibernate native applications currently have to be (manually) restarted when the TM
+    // service restarts, as native applications do not have WildFly service dependencies set for them.
+    private static volatile TransactionSynchronizationRegistry transactionSynchronizationRegistry;
 
+    // JtaSynchronizationStrategy
     @Override
-    protected JtaSynchronizationStrategy getSynchronizationStrategy() {
-        final TransactionSynchronizationRegistry transactionSynchronizationRegistry =
-                locateTransactionSynchronizationRegistry();
+    public void registerSynchronization(Synchronization synchronization) {
+        Assert.checkNotNullParam("transactionSynchronizationRegistry", transactionSynchronizationRegistry).
+                registerInterposedSynchronization(Assert.checkNotNullParam("synchronization", synchronization));
+    }
 
-        return new JtaSynchronizationStrategy() {
+    // JtaSynchronizationStrategy
+    @Override
+    public boolean canRegisterSynchronization() {
+        return Assert.checkNotNullParam("transactionSynchronizationRegistry", transactionSynchronizationRegistry).
+                getTransactionStatus() == Status.STATUS_ACTIVE;
+    }
 
+    @Override
+    protected JtaSynchronizationStrategy getSynchronizationStrategy() {
+        return this;
+    }
 
-            @Override
-            public void registerSynchronization(Synchronization synchronization) {
-                transactionSynchronizationRegistry.registerInterposedSynchronization(synchronization);
-            }
+    /**
+     * Hibernate native applications cannot know when the TransactionManager + TransactionSynchronizationRegistry
+     * services are stopped but JPA container managed applications can and will call setTransactionSynchronizationRegistry
+     * with the new (global) TransactionSynchronizationRegistry to use.
+     *
+     * @param tsr
+     */
+    public static void setTransactionSynchronizationRegistry(TransactionSynchronizationRegistry tsr) {
 
-            @Override
-            public boolean canRegisterSynchronization() {
-                return transactionSynchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE;
+        if ((Assert.checkNotNullParam("tsr", tsr)) != transactionSynchronizationRegistry) {
+            synchronized (WildFlyCustomJtaPlatform.class) {
+                if (tsr != transactionSynchronizationRegistry) {
+                    transactionSynchronizationRegistry = tsr;
+                }
             }
-        };
-
-    }
-
-    private TransactionSynchronizationRegistry locateTransactionSynchronizationRegistry() {
```

Comment:
Can this method be restored, implemented as:

```
TransactionSynchronizationRegistry curTsr = transactionSynchronizationRegistry;
return curTsr != null ? curTsr : (TransactionSynchronizationRegistry) jndiService().locate(TSR_NAME);
```

That achieves the desired perf benefit without taking the risk of bypassing the current mechanism in favor of something else.

The setTransactionSynchronizationRegistry method is only called from HibernatePersistenceProviderAdapater.injectJtaManager which looks to only be called from deep logic within some JPA related DUPs.  If that logic isn't executed (which seems possible, e.g. if there are no JPA deployments) or if some small future change results in it not being set in some scenario, then this static field will not be set.

Reply:
+1 I agree with this.

Reply:
I'm puzzled how I missed this, as I thought I was using the JNDI lookup if the first deployment was native Hibernate ORM.  Thanks for pointing this bug out!